pkg/kubelet/secret/secret_manager.go

```diff
@@ -0,0 +1,275 @@
+	UnregisterPod(pod *v1.Pod)
+}
+
+// simpleSecretManager implements SecretManager interfaces with
+// simple operations to apiserver.
+type simpleSecretManager struct {
+	kubeClient clientset.Interface
+}
+
+func NewSimpleSecretManager(kubeClient clientset.Interface) (Manager, error) {
+	return &simpleSecretManager{kubeClient: kubeClient}, nil
+}
+
+func (s *simpleSecretManager) GetSecret(namespace, name string) (*v1.Secret, error) {
+	return s.kubeClient.Core().Secrets(namespace).Get(name, metav1.GetOptions{})
+}
+
+func (s *simpleSecretManager) RegisterPod(pod *v1.Pod) {
+}
+
+func (s *simpleSecretManager) UnregisterPod(pod *v1.Pod) {
+}
+
+type objectKey struct {
+	namespace string
+	name      string
+}
+
+// secretStoreItems is a single item stored in secretStore.
+type secretStoreItem struct {
+	refCount int
+	secret   *secretData
+}
+
+type secretData struct {
+	sync.Mutex
+	secret         *v1.Secret
+	err            error
+	lastUpdateTime time.Time
+}
+
+/*func (s *secretData) initialized() bool {
+	s.Lock()
+	defer s.Unlock()
+	return s.secret != nil || s.err != nil
+}*/
+
+// update assumes that the secret and error passed to this function
+// we really want to store (e.g. it is not a transient error).
+// Thus, the only case when we want to ignore it is when the passed
+// secret is older then the one already stored.
+func (s *secretData) update(secret *v1.Secret, err error, timestamp time.Time) {
+	s.Lock()
+	defer s.Unlock()
+	if s.secret != nil && secret != nil && isSecretOlder(secret, s.secret) {
+		return
+	}
+	s.secret = secret
+	s.err = err
+	s.lastUpdateTime = timestamp
+}
+
+// secretStore is a local cache of secrets.
+type secretStore struct {
+	kubeClient clientset.Interface
+
+	lock  sync.Mutex
+	items map[objectKey]*secretStoreItem
+	ttl   time.Duration
+}
+
+func newSecretStore(kubeClient clientset.Interface, ttl time.Duration) *secretStore {
+	return &secretStore{
+		kubeClient: kubeClient,
+		items:      make(map[objectKey]*secretStoreItem),
+		ttl:        ttl,
+	}
+}
+
+func isSecretOlder(newSecret, oldSecret *v1.Secret) bool {
+	if newSecret != nil && oldSecret != nil {
+		newVersion, _ := storageetcd.Versioner.ObjectResourceVersion(newSecret)
+		oldVersion, _ := storageetcd.Versioner.ObjectResourceVersion(oldSecret)
+		if newVersion >= oldVersion {
+			return false
+		}
+	}
+	return true
+}
+
+func (s *secretStore) Add(namespace, name string) {
+	key := objectKey{namespace: namespace, name: name}
+
+	// Add is called from RegisterPod, thus it needs to be efficient.
+	// Thus Add() is only increasing refCount and generation of a given secret.
+	// Then Get() is responsible for fetching if needed.
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	item, exists := s.items[key]
+	if !exists {
+		item = &secretStoreItem{
+			refCount: 0,
+			secret:   &secretData{},
+		}
+		s.items[key] = item
+	}
+
+	item.refCount++
+	// This should trigger fetch on the next Get() operation.
+	item.secret.update(nil, nil, time.Now().Add(-s.ttl))
```

Comment:
I thought about it while writing this code and decided to do like this. My motivation was that:
- I made the locking on "item.secret.Lock()" pretty cheap - this is only doing some simple comparison and/or updating and getting values - it should be super cheap
- this really simplifies the code

The problem with making this nil is here is that we would have to handle nil in all other places in this code, which makes the code much less readable and longer. If you really think this should be changed, I can do it, but I think it's not worth it. 
I added an explicit comment that operations holding this mutex should be super fast.